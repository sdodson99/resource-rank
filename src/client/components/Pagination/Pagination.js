import React from 'react';
import PropTypes from 'prop-types';
import styles from './Pagination.module.css';
import coerceNumberToRange from './util/coerceNumberToRange';
import Page from './Page/Page';
import PageGap from './PageGap/PageGap';
import PageRange from './PageRange/PageRange';

const Pagination = ({
  selectedPage,
  pageCount,
  onPageClick,
  selectedPageRange,
  selectedPageGapMinRange,
}) => {
  const firstPage = 1;
  const lastPage = pageCount;
  const coercedSelectedPage = coerceNumberToRange(
    selectedPage,
    firstPage,
    lastPage
  );

  const preSelectedFirstPage = coercedSelectedPage - selectedPageRange;
  const postSelectedLastPage = coercedSelectedPage + selectedPageRange;

  const showPreSelectedPageGap =
    preSelectedFirstPage - firstPage >= selectedPageGapMinRange;
  const showPostSelectedPageGap =
    lastPage - postSelectedLastPage >= selectedPageGapMinRange;

  return (
    <div className={styles.Pagination} data-testid="Pagination">
      {showPreSelectedPageGap && (
        <>
          <Page number={firstPage} onClick={onPageClick} />
          <PageGap />
          <PageRange
            first={preSelectedFirstPage}
            last={coercedSelectedPage - 1}
            onClick={onPageClick}
          />
        </>
      )}
      {!showPreSelectedPageGap && (
        <PageRange
          first={firstPage}
          last={coercedSelectedPage - 1}
          onClick={onPageClick}
        />
      )}

      <Page number={coercedSelectedPage} isSelected={true} />

      {showPostSelectedPageGap && (
        <>
          <PageRange
            first={coercedSelectedPage + 1}
            last={postSelectedLastPage}
            onClick={onPageClick}
          />
          <PageGap />
          <Page number={lastPage} onClick={onPageClick} />
        </>
      )}
      {!showPostSelectedPageGap && (
        <PageRange
          first={coercedSelectedPage + 1}
          last={lastPage}
          onClick={onPageClick}
        />
      )}
    </div>
  );
};

Pagination.propTypes = {
  selectedPage: PropTypes.number,
  pageCount: PropTypes.number,
  onPageClick: PropTypes.func,
  selectedPageRange: PropTypes.number,
  selectedPageGapMinRange: PropTypes.number,
};

Pagination.defaultProps = {
  selectedPage: 1,
  pageCount: 1,
  selectedPageRange: 2,
  selectedPageGapMinRange: 3,
};

export default Pagination;
