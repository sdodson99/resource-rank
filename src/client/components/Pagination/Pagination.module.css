.Pagination {
  @apply flex flex-wrap;
}
