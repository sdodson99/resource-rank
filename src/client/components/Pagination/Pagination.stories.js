import React from 'react';
import Pagination from './Pagination';

export default {
  component: Pagination,
  title: 'Components/Pagination',
};

const Template = (args) => <Pagination {...args} />;

export const Default = Template.bind({});
Default.args = {
  pageCount: 5,
};

export const OnePage = Template.bind({});
OnePage.args = {
  pageCount: 1,
};

export const ManyPages = Template.bind({});
ManyPages.args = {
  pageCount: 20,
};

export const MiddleSelection = Template.bind({});
MiddleSelection.args = {
  selectedPage: 8,
  pageCount: 20,
};

export const SelectedPageRange = Template.bind({});
SelectedPageRange.args = {
  selectedPage: 10,
  pageCount: 20,
  selectedPageRange: 0,
};

export const SelectedPageGapRange = Template.bind({});
SelectedPageGapRange.args = {
  selectedPage: 5,
  pageCount: 10,
  selectedPageRange: 3,
  selectedPageGapMinRange: 5,
};
