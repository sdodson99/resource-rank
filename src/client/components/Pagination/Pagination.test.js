import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createRenderer } from 'react-test-renderer/shallow';
import Pagination from './Pagination';

describe('<Pagination />', () => {
  it('should mount', () => {
    render(<Pagination />);

    const pagination = screen.getByTestId('Pagination');

    expect(pagination).toBeInTheDocument();
  });

  it('should render correctly with defaults', () => {
    const component = createRenderer().render(<Pagination />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when no page gaps needed', () => {
    const component = createRenderer().render(
      <Pagination selectedPage={3} pageCount={5} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when pre-selected page gap needed', () => {
    const component = createRenderer().render(
      <Pagination selectedPage={18} pageCount={20} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when post-selected page gap needed', () => {
    const component = createRenderer().render(
      <Pagination selectedPage={3} pageCount={20} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when page gaps needed, but min gap range overriden', () => {
    const component = createRenderer().render(
      <Pagination
        selectedPage={10}
        pageCount={20}
        selectedPageGapMinRange={10}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when gaps needed and selected page range configured', () => {
    const component = createRenderer().render(
      <Pagination selectedPage={10} selectedPageRange={1} pageCount={20} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with selected page', () => {
    const component = createRenderer().render(
      <Pagination selectedPage={3} pageCount={5} />
    );

    expect(component).toMatchSnapshot();
  });
});
