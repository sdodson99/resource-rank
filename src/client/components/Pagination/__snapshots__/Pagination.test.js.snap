// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render correctly when gaps needed and selected page range configured 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <React.Fragment>
    <Page
      number={1}
    />
    <PageGap />
    <PageRange
      first={9}
      last={9}
    />
  </React.Fragment>
  <Page
    isSelected={true}
    number={10}
  />
  <React.Fragment>
    <PageRange
      first={11}
      last={11}
    />
    <PageGap />
    <Page
      number={20}
    />
  </React.Fragment>
</div>
`;

exports[`<Pagination /> should render correctly when no page gaps needed 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <PageRange
    first={1}
    last={2}
  />
  <Page
    isSelected={true}
    number={3}
  />
  <PageRange
    first={4}
    last={5}
  />
</div>
`;

exports[`<Pagination /> should render correctly when page gaps needed, but min gap range overriden 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <PageRange
    first={1}
    last={9}
  />
  <Page
    isSelected={true}
    number={10}
  />
  <PageRange
    first={11}
    last={20}
  />
</div>
`;

exports[`<Pagination /> should render correctly when post-selected page gap needed 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <PageRange
    first={1}
    last={2}
  />
  <Page
    isSelected={true}
    number={3}
  />
  <React.Fragment>
    <PageRange
      first={4}
      last={5}
    />
    <PageGap />
    <Page
      number={20}
    />
  </React.Fragment>
</div>
`;

exports[`<Pagination /> should render correctly when pre-selected page gap needed 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <React.Fragment>
    <Page
      number={1}
    />
    <PageGap />
    <PageRange
      first={16}
      last={17}
    />
  </React.Fragment>
  <Page
    isSelected={true}
    number={18}
  />
  <PageRange
    first={19}
    last={20}
  />
</div>
`;

exports[`<Pagination /> should render correctly with defaults 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <PageRange
    first={1}
    last={0}
  />
  <Page
    isSelected={true}
    number={1}
  />
  <PageRange
    first={2}
    last={1}
  />
</div>
`;

exports[`<Pagination /> should render correctly with selected page 1`] = `
<div
  className="Pagination"
  data-testid="Pagination"
>
  <PageRange
    first={1}
    last={2}
  />
  <Page
    isSelected={true}
    number={3}
  />
  <PageRange
    first={4}
    last={5}
  />
</div>
`;
